diff --git a/post-receive-email b/post-receive-email
index 8747b84..293cc72 100644
--- a/post-receive-email
+++ b/post-receive-email
@@ -212,7 +212,7 @@ generate_email()
 		describe=$rev
 	fi
 
-	generate_email_header
+	generate_email_header "$1"
 
 	# Call the correct body generation function
 	fn_name=general
@@ -230,17 +230,21 @@ generate_email()
 	else
 		generate_${change_type}_${fn_name}_email | limit_lines $maxlines
 	fi
-
-	generate_email_footer
 }
 
 generate_email_header()
 {
 	# --- Email (all stdout will be the email)
 	# Generate header
+	echo "To: $recipients"
+	if [ -z "$1" ]; then
+		echo "From: $1"
+	fi
+	if [ -z "$envelopesender" ]; then
+		echo "Sender: $envelopesender"
+	fi
 	cat <<-EOF
-	To: $recipients
-	Subject: ${emailprefix}$projectdesc $refname_type $short_refname ${change_type}d. $describe
+	Subject: ${emailprefix} $refname_type $short_refname ${change_type}d. $describe
 	MIME-Version: 1.0
 	Content-Type: text/plain; charset=utf-8
 	Content-Transfer-Encoding: 8bit
@@ -250,11 +254,6 @@ generate_email_header()
 	X-Git-Newrev: $newrev
 	Auto-Submitted: auto-generated
 
-	This is an automated email from the git hooks/post-receive script. It was
-	generated because a ref change was pushed to the repository containing
-	the project "$projectdesc".
-
-	The $refname_type, $short_refname has been ${change_type}d
 	EOF
 }
 
@@ -736,7 +735,7 @@ fi
 recipients=$(git config hooks.mailinglist)
 announcerecipients=$(git config hooks.announcelist)
 envelopesender=$(git config hooks.envelopesender)
-emailprefix=$(git config hooks.emailprefix || echo '[SCM] ')
+emailprefix=$(git config hooks.emailprefix || (basename $(cd $GIT_DIR && pwd) | sed -e s,.git$,,))
 custom_showrev=$(git config hooks.showrev)
 maxlines=$(git config hooks.emailmaxlines)
 diffopts=$(git config hooks.diffopts)
@@ -754,6 +753,6 @@ else
 	while read oldrev newrev refname
 	do
 		prep_for_email $oldrev $newrev $refname || continue
-		generate_email $maxlines | send_mail
+		generate_email $maxlines "$1" | send_mail
 	done
 fi
